### ISSUE TITLE
Sound is leaking across multiple xpra sessions attached to different machines.

### ISSUE THREAD

**Issue by totaam** [Xpra-org/xpra#1266]
*Issue migrated from trac ticket # 1266*

**component:** sound | **priority:** major | **resolution:** fixed

#### 2016-07-24 05:33:44: kundanvp created the issue

___
> If we attach two xpra clients from two different machines, attached to two separate xpra server on one machine. (Both xpra server has xterm as it's child application)
> 
> And, if we start any sound playing application, like vlc player on any one of the xpra xterm session attached to one of the xpra server, we see sound coming from the other machine as well.
> 
> However, the expectation is that, since we have opened the vlc player from one xpra client machine only, it should not expose it's sound to the other xpra client machine.
> 
> This seems to be a regression from the behaviour of xpra v0.16 version.
> 
> Minimum configuration needed to reproduce: 3 Ubuntu 16.04 machines.
> 
> Steps to reproduce:-
> 
> 1. Start two xpra server on first machine, which is acting like server and don't have existing pulseaudio server running.
> 
> ```
> xpra start :100 --start-child=xterm --no-daemon --bind-tcp=192.168.56.1:10000
> xpra start :101 --start-child=xterm --no-daemon --bind-tcp=192.168.56.1:10001
> ```
> 
> 2. Attach with 1st xpra server from second machine with command
> 
> ```
> xpra attach tcp:192.168.56.1:10000
> ```
> 
> 3. Attach with 2nd xpra server from second machine with command
> 
> ```
> xpra attach tcp:192.168.56.1:10001
> ```
> 
> 4. Open a sound playing application let's say vlc from 3rd machine xterm session which has got exposed through xpra attach.
> 
> 5. Play a sound file in vlc.
> 
> 6. We see sound coming from the 2nd machine as well.
> 
> 
> 
>

**Comment by totaam** [comment]
It's very unlikely that this ever worked properly, even with previous versions unless each session is started from a different user: you cannot start more than one pulseaudio server per user without some major hacks.

So when you start your second xpra instance, the pulseaudio server will fail to start and the session will end up re-using the pulseaudio server started for the first session.

You can verify this by posting your server log file, which should contain the message: *Warning: pulseaudio has terminated shortly after startup..*

**Comment by totaam** [comment]
Oh. I didn't know that.
So Debian or Ubuntu must be patching pulseaudio to do that as my previous requests to pulseaudio developers were waved away.

If so, then the changes in #1141 may have caused this: it is possible that when we call pactl to set the default device it somehow affects both servers?

I don't have Ubuntu or Debian at hand to test.

**Comment by totaam** [comment]
The `XDG_RUNTIME_DIR` hack is what I was talking about in comment:1, that's the code I had written for winswitch and I'm not sure we want to use the same hack here: I would be very reluctant to modify this environment variable for all applications as it may have undesirable side effects (see #1129 #1105) when we only want pulseaudio to behave.

Can other applications still locate the pulseaudio server if `XDG_RUNTIME_DIR` is changed only for pulseaudio? I doubt it, how would they? X11 applications may be able to use the X11 display environment variables (see `xprop -root | grep -i pulse`, but command line applications would not.

**Comment by totaam** [comment]
On Fedora and centos, it is impossible to start another pulseaudio instance if one is running already, no matter what is set for the unix domain socket path or XDG_RUNTIME_DIR. This is what you get if you try:
* with `--load=module-native-protocol-unix socket=$HOME/tmp"`:
```
N: [pulseaudio] main.c: User-configured server at {0d37226ca0064aaab1db9e66016c45f5}unix:/run/user/1000/pulse/native, which appears to be local. Probing deeper.
E: [pulseaudio] pid.c: Daemon already running.
```
* with `XDG_RUNTIME_DIR=`pwd`/tmp/`:
```
N: [pulseaudio] main.c: User-configured server at {0d37226ca0064aaab1db9e66016c45f5}unix:/run/user/1000/pulse/native, refusing to start/autospawn.
```

If a workaround is possible, it won't work on Fedora or centos, so I am not very interested in working on this, sorry. Patches welcome though.
The "pulseaudio-command" option should expand environment variables, if you want to use that. ie: `$DISPLAY` should be set by the time we start pulseaudio.

**Comment by totaam** [comment]
Another Ubuntu-only? sound issue (see also #1141 and #1219), I don't have time for Ubuntu - sorry. Re-scheduling.

**Comment by totaam** [comment]
@maxmylyn: if your setup only ever uses one instance per user account, you should disable this feature as it may cause problems in some corner cases.

**Comment by totaam** [comment]
Noted - I'll pass this along.

**Comment by totaam** [comment]
Related improvements (loop detection, notifications, etc): [r18468](../commit/c8fce11d3951c4c07511377760b91fcbaf7c7718)

**Comment by totaam** [comment]
[r19064](../commit/7d19ad19da7a165f1a028d6ea1c9722d1b38da85)+[r19065](../commit/da9218480e3d670335291562277a1491d3032d9b) is a large-ish fixup for [r18103](../commit/ed1db4712a0b3822ef5d557263faba8855c25c00) (see comment:12) because memfd is not available on Ubuntu 16.04.

Follow up: #2186
